server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -301,52 +276,57 @@ private Criteria createModifiesDateFilteringCriteria(Set<Consumer> consumers, Da
-        // overlapping entitlements that provide them
-        ProductEntitlements pidEnts = getOverlappingForModifying(entitlements);
-        if (pidEnts.isEmpty()) {
-            // Empty collections break hibernate queries
-            return modifying;
-        }
-
-        List<Product> overlappingProducts = productCurator.listAllByIds(pidEnts.getAllProductIds());
-
-        Set<String> entitlementProducts = new HashSet<String>();
-        for (Entitlement entitlement : entitlements) {
-            entitlementProducts.add(entitlement.getPool().getProductId());
-            for (Product product : entitlement.getPool().getProvidedProducts()) {
-                entitlementProducts.add(product.getId());
-            }
-        }
-
-        for (Product overlappingProduct : overlappingProducts) {
-            boolean modifies = false;
-            Iterator<String> ppit = entitlementProducts.iterator();
-            // No need to continue checking once we have found a modified product
-            while (!modifies && ppit.hasNext()) {
-                modifies = overlappingProduct.modifies(ppit.next());
-            }
-            if (modifies) {
-                // Return all entitlements for the modified product
-                modifying.addAll(pidEnts.getEntitlementsByProductId(overlappingProduct.getId()));
+    public Collection<String> batchListModifying(Iterable<Entitlement> entitlements) {
+        List<String> eids = new LinkedList<String>();
+
+        if (entitlements != null && entitlements.iterator().hasNext()) {
+            String hql =
+                "SELECT DISTINCT eOut.id" +
+                "    FROM Entitlement eOut" +
+                "        JOIN eOut.pool outPool" +
+                "        JOIN outPool.providedProducts outProvided" +
+                "        JOIN outProvided.productContent outProvContent" +
+                "        JOIN outProvContent.content outContent" +
+                "        JOIN outContent.modifiedProductIds outModProdId" +
+                "    WHERE" +
+                "        outPool.endDate >= current_date AND" +
+                "        eOut NOT IN (:ein) AND" +
+                "        EXISTS (" +
+                "            SELECT eIn" +
+                "                FROM Entitlement eIn" +
+                "                    JOIN eIn.consumer inConsumer" +
+                "                    JOIN eIn.pool inPool" +
+                "                    JOIN inPool.product inMktProd" +
+                "                    LEFT JOIN inPool.providedProducts inProvidedProd" +
+                "                WHERE eIn in (:ein) AND inConsumer = eOut.consumer AND" +
+                "                    inPool.endDate >= outPool.startDate AND" +
+                "                    inPool.startDate <= outPool.endDate AND" +
+                "                    (inProvidedProd.id = outModProdId OR inMktProd.id = outModProdId)" +
+                "        )";
```

Comment:
Might be worth adding a description here of what this query is actually doing. Might be a little hard to follow for someone that isn't that strong with reading SQL queries.